File: core/java/src/main/java/hara/lang/protocol/IContextLifeCycle.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.icontextlifecycle", name = "IContextLifeCycle")
public interface IContextLifeCycle {
  @HaraMethod(value = "has-module?", arity = 2)
  boolean hasModule(Object moduleId);

  @HaraMethod(value = "setup-module", arity = 2)
  void setupModule(Object moduleId);

  @HaraMethod(value = "teardown-module", arity = 2)
  void teardownModule(Object moduleId);

  @HaraMethod(value = "has-setup-ptr?", arity = 2)
  boolean hasSetupPtr(IPointer pointer);

  @HaraMethod(value = "setup-ptr", arity = 2)
  void setupPtr(IPointer pointer);

  @HaraMethod(value = "teardown-ptr", arity = 2)
  void teardownPtr(IPointer pointer);
}

File: core/rust/src/lang/protocol/icontextlifecycle.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.icontextlifecycle",
    name = "IContextLifeCycle"
)]
pub trait IContextLifeCycle<M, P> {
    #[hara_method(value = "has-module?", arity = 2)]
    fn has_module(&self, module: &M) -> bool;
    #[hara_method(value = "setup-module", arity = 2)]
    fn setup_module(&mut self, module: M);
    #[hara_method(value = "teardown-module", arity = 2)]
    fn teardown_module(&mut self, module: &M);
    #[hara_method(value = "has-setup-ptr?", arity = 2)]
    fn has_setup_ptr(&self, pointer: &P) -> bool;
    #[hara_method(value = "setup-ptr", arity = 2)]
    fn setup_ptr(&mut self, pointer: P);
    #[hara_method(value = "teardown-ptr", arity = 2)]
    fn teardown_ptr(&mut self, pointer: &P);
}
